[PageImpl] Non-deterministic behavior or `HashSet` might fail tests in `PageImplTest`

<!-- 

Before sending a bug report please check if the latest release (if you're not already using it) fixes your problem. Alternatively, you could
install the latest snapshot version from the master branch.
 -->

## Bug Report

**Current Behavior**
- While testing the `PageImpl`, the contents of `Page` are mocked using `MockPage` from the dependency `io.wcm.testing.aem-mock.junit5`.
- The mock values for `Page` is loaded from the following files based on the version of the `PageImplTest`, where the `tags` are in the order `[one, two, three]`.
  - `src/test/resources/page/test-content.json`
  - `src/test/resources/page/v2/test-content.json`
  - `src/test/resources/page/v3/test-content.json`
- The mocking occurs after instantiation of `PageImpl` at https://github.com/adobe/aem-core-wcm-components/blob/3524e33424bf6831faca014284c5bad803d3744d/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageImpl.java#L132-L137
- However, while mocking the `Tags` of `PageImpl` at `MockTagManager:380` (screenshots provided below), the values of the tags are converted from the loaded resource into a `HashSet`.
- The official [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/HashSet.html) mentions that `HashSet` does not guarantee that the order will be maintained over time.
- This causes issues where the order of tags of tags is expected to be in a specific order.
- Specifically, the following tests expects the order of the tags after exporting to JSON to be in the order ["three", "one", "two"]
  - `com.adobe.cq.wcm.core.components.internal.models.v1.PageImplTest.testPage`
  - `com.adobe.cq.wcm.core.components.internal.models.v2.PageImplTest.testPage`
  - `com.adobe.cq.wcm.core.components.internal.models.v2.PageImplTest.testPageWithDeprecatedCaconfig`
  - `com.adobe.cq.wcm.core.components.internal.models.v3.PageImplTest.testPage`
  - `com.adobe.cq.wcm.core.components.internal.models.v3.PageImplTest.testPageWithDeprecatedCaconfig`
#### Steps to reproduce
- The above mentioned issue is found using the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool, which shuffles the order of `HashSet`.
```
# Clone the Repo
https://github.com/adobe/aem-core-wcm-components.git

# Compile the module byte-buddy-dep
mvn clean install -pl bundles/core -am -DskipTests

# (Optional) Run the unit test
mvn -pl bundles/core test -Dtest=<Above Mentioned Test Name>

# Run the unit test using NonDex
mvn -pl bundles/core edu.illinois:nondex-maven-plugin:2.1.1:nondex -Dtest=<Above Mentioned Test Name>
```
- Following the above steps for the mentioned tests will produce the following result:
```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   PageImplTest.testPage:109->testPage:142 \
expected: <{"id":"page-bb590de134","designPath":"/etc/designs/mysite","title":"Templated Page","description":"Description","brandSlug":"Brand Slug","lastModifiedDate":1453282416000,"templateName":"product-page","appResourcesPath":"/core/etc.clientlibs/wcm/core/page/clientlibs/favicon/resources","htmlPageItems":[{"element":"link","location":"header","attributes":{"href":"/_theme/theme.css","rel":"preload","as":"style"}},{"element":"script","location":"footer","attributes":{"async":true,"defer":false,"src":"/_theme/theme.js"}},{"element":"meta","location":"header","attributes":{"charset":"UTF-8"}}],"cssClassNames":"class1 class2","language":"en-GB",":itemsOrder":["root"],":items":{"root":{":itemsOrder":["text"],":items":{"text":{"id":"text-41bcd02e64","text":"<p>This is a test</p>","richText":true,":type":"core/wcm/components/text/v2/text","dataLayer":{"text-41bcd02e64":{"@type":"core/wcm/components/text/v2/text","repo:modifyDate":"2016-01-18T15:59:43Z","xdm:text":"<p>This is a test</p>"}}}},":type":"wcm/foundation/components/responsivegrid"}},":type":"core/wcm/components/page/v2/page","dataLayer":{"page-bb590de134":{"@type":"core/wcm/components/page/v2/page","repo:modifyDate":"2016-01-20T09:33:36Z","dc:title":"Templated Page","dc:description":"Description","xdm:template":"/conf/coretest/settings/wcm/templates/product-page","xdm:language":"en-GB","xdm:tags":["three","one","two"],"repo:path":"/core/content/page/templated-page.html"}}}> \
 but was: <{"id":"page-bb590de134","lastModifiedDate":1453282416000,"brandSlug":"Brand Slug","description":"Description","title":"Templated Page","designPath":"/etc/designs/mysite","templateName":"product-page","htmlPageItems":[{"attributes":{"href":"/_theme/theme.css","rel":"preload","as":"style"},"element":"link","location":"header"},{"attributes":{"async":true,"defer":false,"src":"/_theme/theme.js"},"element":"script","location":"footer"},{"attributes":{"charset":"UTF-8"},"element":"meta","location":"header"}],"appResourcesPath":"/core/etc.clientlibs/wcm/core/page/clientlibs/favicon/resources","cssClassNames":"class1 class2","language":"en-GB",":items":{"root":{":itemsOrder":["text"],":items":{"text":{"id":"text-41bcd02e64","text":"<p>This is a test</p>","richText":true,":type":"core/wcm/components/text/v2/text","dataLayer":{"text-41bcd02e64":{"repo:modifyDate":"2016-01-18T15:59:43Z","@type":"core/wcm/components/text/v2/text","xdm:text":"<p>This is a test</p>"}}}},":type":"wcm/foundation/components/responsivegrid"}},":itemsOrder":["root"],":type":"core/wcm/components/page/v2/page","dataLayer":{"page-bb590de134":{"repo:modifyDate":"2016-01-20T09:33:36Z","dc:title":"Templated Page","@type":"core/wcm/components/page/v2/page","dc:description":"Description","xdm:language":"en-GB","xdm:template":"/conf/coretest/settings/wcm/templates/product-page","xdm:tags":["one","three","two"],"repo:path":"/core/content/page/templated-page.html"}}}>
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
```
- Filtering out the tags from the above log
```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   PageImplTest.testPage:109->testPage:142 
expected: ..."xdm:tags":["three","one","two"]... 
  but was: ..."xdm:tags":["one","three","two"]...
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
```

**Expected behavior/code**
- The ordering of the tags should have been maintain in a specific order for all runs 
- The tests should have passed successfully when run through `NonDex`

**Environment**
- Core Components version: `2.23.5-SNAPSHOT`
- Java version: `Java(TM) SE Runtime Environment (build 1.8.0_381-b09)`

**Possible Solution**
<!--- If you have suggestions for the bug fix. Please also consider creating a Pull Request 🤘. -->
- Even though the order of key value pair could change in the exported JSON, the order of an array needs to be maintained. So, changing the test function to compare based on the presence of elements in an array (like a Set) is incorrect.
- 1st Possible Solution
  - Change the implementation of the dependency `MockTagManager` to use `LinkedHashSet` instead of `HashSet`
  - This ensures that the order of tags are maintained for all the runs.
  - However, this solution requires changing code out of the project scope.
```
+/-  io.wcm.testing.mock.aem.MockTagManager.java:380
-    Set<Tag> treeTags = new HashSet<>();
+    Set<Tag> treeTags = new LinkedHashSet<>();
```
  - 2nd Possible Solution
    - Sort the ordering of the tags in `PageImpl` based on the tag name. This ensures that whatever may be order returned by the `MockTagManager`, we would maintain the same order while exporting to JSON.
```
+/-  com.adobe.cq.wcm.core.components.internal.models.v1.PageImpl.java:134

    Tag[] tags = currentPage.getTags();
    keywords = new String[tags.length];

    Tag[] tags = currentPage.getTags();
+   Arrays.sort(tags, Comparator.comparing(Tag::getName));
    keywords = new String[tags.length];
```
 - However, this also requires to change the expected order of exported tags from `["three", "one", "two"]` to ["one", "three", "two"] because of lexicographic order sorting.
  - The ordering of tags needs to changed in the following expected JSON files
      -  `src/test/resources/page/exporter-templated-page.json`
      - `src/test/resources/page/v3/exporter-templated-page.json` 
      - `src/test/resources/page/v3/exporter-templated-page.json`

```
      "xdm:tags": [
-       "three",
-       "one",
        "two"
      ]

      "xdm:tags": [
+       "one",
+       "three",
        "two"
      ]
```
- **If the proposed solution looks good, I can create a PR based on your review.**

**Additional context / Screenshots**
- The order of tags loaded from the resource
<img width="876" alt="Screenshot 2023-11-02 at 7 33 38 PM" src="https://github.com/adobe/aem-core-wcm-components/assets/46614118/47590701-add5-4461-9054-bb3ca4584d35">

- The first type of ordering of the tags when inserted into the `HashSet` is `["three",  "one", "two"]`
<img width="715" alt="Screenshot 2023-11-02 at 7 34 39 PM" src="https://github.com/adobe/aem-core-wcm-components/assets/46614118/5aed4978-b951-4c1c-8831-5f2e57b0746a">

- The second type of ordering of the tags when run with `Nondex` is `["one",  "three", "two"]`
<img width="740" alt="Screenshot 2023-11-02 at 7 38 40 PM" src="https://github.com/adobe/aem-core-wcm-components/assets/46614118/b64d5c93-da90-463c-86b5-473e7815bb14">

- The above incorrect order of tags is transformed to an array of tags
<img width="900" alt="Screenshot 2023-11-02 at 7 39 01 PM" src="https://github.com/adobe/aem-core-wcm-components/assets/46614118/896b3a57-97d4-4012-b096-ae0924ad483a">

- `keyword` array value is also affected, since it depends on the order of `tags`
<img width="587" alt="Screenshot 2023-11-02 at 7 39 34 PM" src="https://github.com/adobe/aem-core-wcm-components/assets/46614118/a986dcbe-a31a-4b93-93e0-f3821c449a32">

--- a/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageImpl.java
+++ b/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageImpl.java
@@ -131,6 +131,7 @@ public class PageImpl extends AbstractComponentImpl implements Page {
             title = currentPage.getName();
         }
         Tag[] tags = currentPage.getTags();
+        // sort to maintain the order of tag elements
         Arrays.sort(tags, Comparator.comparing(Tag::getName));
         keywords = new String[tags.length];
         int index = 0;

--- a/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageImpl.java
+++ b/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageImpl.java
@@ -19,6 +19,7 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Calendar;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -130,6 +131,7 @@ public class PageImpl extends AbstractComponentImpl implements Page {
             title = currentPage.getName();
         }
         Tag[] tags = currentPage.getTags();
+        Arrays.sort(tags, Comparator.comparing(Tag::getName));
         keywords = new String[tags.length];
         int index = 0;
         for (Tag tag : tags) {

--- a/bundles/core/src/test/resources/page/exporter-templated-page.json
+++ b/bundles/core/src/test/resources/page/exporter-templated-page.json
@@ -18,8 +18,8 @@
             "xdm:template": "/conf/coretest/settings/wcm/templates/product-page",
             "repo:path": "/core/content/page/templated-page.html",
             "xdm:tags": [
-                "three",
                 "one",
+                "three",
                 "two"
             ]
         }

--- a/bundles/core/src/test/resources/page/v2/exporter-templated-page.json
+++ b/bundles/core/src/test/resources/page/v2/exporter-templated-page.json
@@ -72,8 +72,8 @@
       "xdm:template": "/conf/coretest/settings/wcm/templates/product-page",
       "xdm:language": "en-GB",
       "xdm:tags": [
-        "three",
         "one",
+        "three",
         "two"
       ],
       "repo:path": "/core/content/page/templated-page.html"

--- a/bundles/core/src/test/resources/page/v3/exporter-templated-page.json
+++ b/bundles/core/src/test/resources/page/v3/exporter-templated-page.json
@@ -72,8 +72,8 @@
       "xdm:template": "/conf/coretest/settings/wcm/templates/product-page",
       "xdm:language": "en-GB",
       "xdm:tags": [
-        "three",
         "one",
+        "three",
         "two"
       ],
       "repo:path": "/core/content/page/templated-page.html"

